src/app/ipv/middleware.js

```diff
@@ -319,213 +313,237 @@ async function handleUnexpectedPage(req, res, pageId) {
-    res.render(getIpvPageTemplatePath(PAGES.PYI_ATTEMPT_RECOVERY), {
-      csrfToken: req.csrfToken(),
-    });
-  },
-  updateJourneyState: async (req, res, next) => {
-    try {
-      const currentPageId = req.params.pageId;
-      const { action } = req.params;
-
-      if (action && isValidPage(currentPageId)) {
-        await handleJourneyResponse(req, res, action, currentPageId);
-      } else {
-        res.status(HTTP_STATUS_CODES.NOT_FOUND);
-        return res.render(getTemplatePath("errors", "page-not-found"));
-      }
-    } catch (error) {
-      next(error);
+function render404(response) {
+  response.status(HTTP_STATUS_CODES.NOT_FOUND);
+  response.render(getErrorPageTemplatePath(ERROR_PAGES.PAGE_NOT_FOUND));
+}
+
+function renderTechnicalError(request, response) {
+  request.session.currentPage = PAGES.PYI_TECHNICAL;
+  response.status(HTTP_STATUS_CODES.UNAUTHORIZED);
+  response.render(getIpvPageTemplatePath(PAGES.PYI_TECHNICAL), {
+    context: "unrecoverable",
+  });
+}
+
+async function renderAttemptRecoveryPage(req, res) {
+  res.render(getIpvPageTemplatePath(PAGES.PYI_ATTEMPT_RECOVERY), {
+    csrfToken: req.csrfToken(),
+  });
+}
+
+async function updateJourneyState(req, res, next) {
+  try {
+    const currentPageId = req.params.pageId;
+    const action = req.params.action;
+
+    if (action && isValidIpvPage(currentPageId)) {
+      await handleJourneyResponse(req, res, action, currentPageId);
+    } else {
+      render404(res);
     }
-  },
-  handleJourneyPage: async (req, res, next) => {
-    try {
-      const { pageId } = req.params;
-      const { context } = req?.session || "";
-
-      // handles page id validation first
-      if (!isValidPage(pageId)) {
-        res.status(HTTP_STATUS_CODES.NOT_FOUND);
-        return res.render(getTemplatePath("errors", "page-not-found"));
-      }
+  } catch (error) {
+    next(error);
+  }
+}
 
-      if (!req.session?.ipvSessionId) {
-        logError(
-          req,
-          {
-            pageId: pageId,
-            expectedPage: req.session.currentPage,
-          },
-          "req.ipvSessionId is null",
-        );
-
-        req.session.currentPage = PAGES.PYI_TECHNICAL;
-        return res.render(getIpvPageTemplatePath(req.session.currentPage), {
-          context: "unrecoverable",
-        });
-      } else if (pageId === PAGES.PYI_TIMEOUT_UNRECOVERABLE) {
-        req.session.currentPage = PAGES.PYI_TIMEOUT_UNRECOVERABLE;
-        return res.render(getIpvPageTemplatePath(req.session.currentPage));
-      } else if (req.session.currentPage !== pageId) {
-        return await handleUnexpectedPage(req, res, pageId);
-      }
+async function handleJourneyPage(req, res, next) {
+  return handleJourneyPageInternal(req, res, next);
+}
 
-      const renderOptions = {
-        pageId,
-        csrfToken: req.csrfToken(),
-        context,
-      };
+async function handleJourneyPageInternal(
+  req,
+  res,
+  next,
+  pageErrorState = undefined,
```

Comment:
I'm guessing Express doesn't like it if you put this on `handleJourneyPage` itself? maybe this 'internal' function could be a bit better named something like 'renderJourneyPageWithOptionalError`

Reply:
Oh, that's a good point, I didn't try that